### PR TITLE
IGNITE-21618 In-flights for read-only transactions

### PR DESCRIPTION
**Pull request by denis-chudov**
https://issues.apache.org/jira/browse/IGNITE-21618

### CODE REVIEW

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionInflights.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        @Override
+        public String toString() {
+            return "ReadOnlyTxContext [inflights=" + inflights + ']';
+        }
+    }
+
+    static class ReadWriteTxContext extends TxContext {
+        private final CompletableFuture<Void> waitRepFut = new CompletableFuture<>();
+        private final PlacementDriver placementDriver;
+        volatile CompletableFuture<Void> finishInProgressFuture = null;
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionInflights.java:215]: it is used in TxManager

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java [hunk 2]
```diff
@@ -1848,6 +1878,12 @@ private class PartitionScanSubscription implements Subscription {
 
             private final AtomicLong requestedItemsCnt;
 
+            private final TransactionInflights transactionInflights;
+
+            private boolean readOnlyTransaction;
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java:1883]: no, let's leave as is, I better remove these fields

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java [hunk 3]
```diff
@@ -1856,11 +1892,19 @@ private class PartitionScanSubscription implements Subscription {
...
+                    Subscriber<? super BinaryRow> subscriber,
+                    TransactionInflights transactionInflights,
+                    boolean readOnlyTransaction,
+                    UUID txId
+            ) {
                 this.subscriber = subscriber;
                 this.canceled = new AtomicBoolean(false);
                 this.scanId = CURSOR_ID_GENERATOR.getAndIncrement();
                 this.requestedItemsCnt = new AtomicLong(0);
+                this.transactionInflights = transactionInflights;
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java:1905]: done

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java [hunk 4]
```diff
@@ -1802,6 +1820,12 @@ private static class PartitionScanPublisher implements Publisher<BinaryRow> {
         /** True when the publisher has a subscriber, false otherwise. */
         private final AtomicBoolean subscribed;
 
+        private final TransactionInflights transactionInflights;
+
+        private boolean readOnlyTransaction;
+
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java:1826]: fixed

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionInflights.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionInflights.java:215]: fixed

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionInflights.java [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+
+        abstract void finishTx(Map<TablePartitionId, IgniteBiTuple<ClusterNode, Long>> enlistedGroups);
+
+        abstract boolean isTxFinishing();
+
+        abstract boolean isReadyToFinish();
+    }
+
+    static class ReadOnlyTxContext extends TxContext {
+        volatile boolean markedFinished;
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionInflights.java:184]: fixed

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java [hunk 6]
```diff
@@ -810,6 +786,8 @@ CompletableFuture<Void> completeReadOnlyTransactionFuture(TxIdAndTimestamp txIdA
 
         readOnlyTxFuture.complete(null);
 
+        transactionInflights.markReadOnlyTxFinished(txIdAndTimestamp.getTxId());
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java:789]: done

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/FinishedReadOnlyTransactionTracker.java [hunk 7]
```diff
@@ -70,25 +79,30 @@ public void broadcastClosedTransactions() {
         }
 
         Set<UUID> txToSend = finishedTransactions.stream()
+                .filter(transactionInflights::inflightsCompleted)
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/FinishedReadOnlyTransactionTracker.java]: done

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java [hunk 8]
```diff
@@ -1573,7 +1583,9 @@ private Publisher<BinaryRow> readWriteScan(
                     return replicaSvc.invoke(recipient.node(), request);
                 },
                 // TODO: IGNITE-17666 Close cursor tx finish.
-                (intentionallyClose, fut) -> completeScan(txId, tablePartitionId, fut, recipient.node(), intentionallyClose));
+                (intentionallyClose, fut) -> completeScan(txId, tablePartitionId, fut, recipient.node(), intentionallyClose),
+                new ReadWriteInflightBatchRequestTracker()
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java]: done

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionInflights.java [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+            ctx.removeInflight(txId);
+
+            return ctx;
+        });
+
+        // Avoid completion under lock.
+        tuple.onRemovedInflights();
+    }
+
+    Collection<UUID> readOnlyTxContextsReadyToFinish(int limit) {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionInflights.java]: renamed

**Review by sanpwc — APPROVED** [review]

**Pull request merged**